jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CreateOrUpdateCredentialProfilesAction.kt

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.credentials
+
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.PlatformDataKeys
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.fileEditor.FileEditorManager
+import com.intellij.openapi.fileEditor.OpenFileDescriptor
+import com.intellij.openapi.fileTypes.FileTypes
+import com.intellij.openapi.fileTypes.ex.FileTypeManagerEx
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.Messages
+import com.intellij.openapi.vfs.LocalFileSystem
+import org.jetbrains.annotations.TestOnly
+import software.amazon.awssdk.profiles.ProfileFileLocation
+import software.aws.toolkits.core.utils.exists
+import software.aws.toolkits.resources.message
+import software.aws.toolkits.telemetry.AwsTelemetry
+import java.nio.file.Path
+
+class CreateOrUpdateCredentialProfilesAction @TestOnly constructor(
+    private val writer: ConfigFilesFacade
+) : AnAction(message("configure.toolkit.upsert_credentials.action")), DumbAware {
+    @Suppress("unused")
+    constructor() : this(
+        DefaultConfigFilesFacade(
+            configPath = ProfileFileLocation.configurationFilePath(),
+            credentialsPath = ProfileFileLocation.credentialsFilePath()
+        )
+    )
+
+    private val localFileSystem = LocalFileSystem.getInstance()
+
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.getRequiredData(PlatformDataKeys.PROJECT)
+
+        // if both config and credential files do not exist, create a new config file
+        if (!writer.configPath.exists() && !writer.credentialsPath.exists()) {
+            if (confirm(project, writer.configPath)) {
+                try {
+                    writer.createConfigFile()
+                } finally {
+                    AwsTelemetry.createCredentials(project)
+                }
+            } else {
+                return
+            }
+        }
+
+        // open both config and credential files, if they exist
+        // credential file is opened last since it takes precedence over the config file
+        val virtualFiles = listOf(writer.configPath.toFile(), writer.credentialsPath.toFile()).filter { it.exists() }.map {
+            localFileSystem.refreshAndFindFileByIoFile(it) ?: throw RuntimeException(
+                message(
+                    "credentials.could_not_open",
+                    it
+                )
+            )
+        }
+
+        val fileEditorManager = FileEditorManager.getInstance(project)
+
+        localFileSystem.refreshFiles(virtualFiles, false, false) {
+            virtualFiles.forEach {
+                if (it.fileType == FileTypes.UNKNOWN) {
+                    ApplicationManager.getApplication().runWriteAction {
```

Comment:
Is the `runWriteAction` required if we are only opening the files?

Reply:
yes